src/filters/common.js

```diff
@@ -0,0 +1,99 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module paste-from-office/filters/common
+ */
+
+/**
+ * Removes paragraph wrapping content inside list item.
+ *
+ * @param {module:engine/view/element~Element|module:engine/view/documentfragment~DocumentFragment} elementOrDocumentFragment
+ * @param {module:engine/view/upcastwriter~UpcastWriter} writer
+ */
+export function unwrapParagraph( elementOrDocumentFragment, writer ) {
+	const iterableNodes = elementOrDocumentFragment.is( 'element' ) ? elementOrDocumentFragment.getChildren() : elementOrDocumentFragment;
+
+	for ( const element of iterableNodes ) {
+		if ( element.is( 'element', 'p' ) && element.parent.is( 'element', 'li' ) ) {
+			unwrapSingleElement( element, writer );
+		}
+
+		if ( element.is( 'element' ) && element.childCount ) {
+			unwrapParagraph( element, writer );
+		}
+	}
+}
+
+/**
+ * Moves nested list inside previous sibling element, what is a proper HTML standard.
+ *
+ * @param {module:engine/view/element~Element|module:engine/view/documentfragment~DocumentFragment} elementOrDocumentFragment
+ * @param {module:engine/view/upcastwriter~UpcastWriter} writer
+ */
+export function moveNestedListToListItem( elementOrDocumentFragment, writer ) {
+	// There are 2 situations which are fixed in the for loop:
+	//
+	// 1. Unwrap nested list to avoid situation that UL or OL is direct child of another UL or OL.
+	// OL                                OL
+	// |-> LI                            |-> LI
+	//     |-> OL                            |-> OL
+	//         |-> OL                            |-> LI
+	//         |   |-> OL                        |-> LI
+	//         |       |-> OL
+	//         |           |-> LI
+	//         |-> LI
+	// or list like:
+	// OL                                OL
+	// |-> OL                            |-> LI
+	//     |-> OL
+	//          |-> OL
+	//              |-> LI
+	//
+	// 2. Move list to previous list item:
+	// OL                                OL
+	// |-> LI                            |-> LI
+	// |-> OL                                |-> OL
+	//     |-> LI                                |-> LI
+	const iterableNodes = elementOrDocumentFragment.is( 'element' ) ? elementOrDocumentFragment.getChildren() : elementOrDocumentFragment;
```

Comment:
I find this method a bit complicated because of the recursion.

I did small tests and it seems that we can get rid of the recursion here by using tree walker (by creating a range inside document fragment).

I'd go with something like this:

```js
for ( const value of writer.createRangeIn( elementOrDocumentFragment ) ) {
		const element = value.item;

		// 2.
		if ( element.is( 'li' ) ) {
			const next = element.nextSibling;

			if ( next && isList( next ) ) {
				writer.remove( next );
				writer.insertChild( element.childCount, next, element );
			}
		}

		// 1.
		if ( isList( element ) ) {
			let firstChild = element.getChild( 0 );

			while ( isList( firstChild ) ) {
				unwrapSingleElement( firstChild, writer );
				firstChild = element.getChild( 0 );
			}
		}
	}
```

I think that using tree walker here should be safe as long we mangle with a node after (not a node before) or it's children of current walker value.